Updating getting extensions to allow for excluding extensions from an option

oikeme opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

Fixes p1HpG7-6RI-p2

#### Changes proposed in this Pull Request:
This updates the final list of extensions retrieved to include whitelisted extensions minus a list of extensions to be excluded from options if the option is set.


#### Testing instructions:
Testing options:

- Add an option for jetpack_extensions_exclusions
- Include extensions to be excluded in the option
- Verify that the extensions excluded in the option are not available
- Run tests for ```test_get_available_extensions``` in  ```tests/php/test_class.jetpack-gutenberg.php```

#### Proposed changelog entry for your changes:
- Adding option to allow exclusion of whitelisted extensions from final extension list.

oikeme committed:
Updating get_jetpack_gutenberg_extensions_whitelist

Updating option for whitelist to include exclusions from wp_options for available blocks

oikeme committed:
Update test_class.jetpack-gutenberg.php

Adding tests for get_available_extensions addition.

oikeme committed:
Update test_class.jetpack-gutenberg.php

Update tests to set available extensions

oikeme committed:
Revert "Update test_class.jetpack-gutenberg.php"

This reverts commit 5fcda9029572274a16652494e6cf5da63c5be8a6.

ebinnion commented on class.jetpack-gutenberg.php:
```diff
@@ -326,6 +326,20 @@ public static function get_jetpack_gutenberg_extensions_whitelist() {
+	 * @return array A list of blocks: eg [ 'publicize', 'markdown' ]
+	 */
+	public static function get_available_extensions() {
+		$exclusions             = get_option( 'jetpack_extensions_exclusions', array() );
```
Since we're adding this option, will you add it to:

- The options whitelist in the `Jetpack_Sync_Defaults` class so that we can sync it and provide some support tools as necessary
- In the `Jetpack_Options::get_all_wp_options()` array so that we can clear the option out with our reset tools

oikeme replied:
Thanks! updated and added to both.

eliorivero commented on class.jetpack-gutenberg.php:
```diff
@@ -326,6 +326,20 @@ public static function get_jetpack_gutenberg_extensions_whitelist() {
+
+		$combined_extension_list = array_diff( $whitelisted_extensions, $exclusions );
+
+		return $combined_extension_list;
```
This could save setting these three redundant variables:

```php
return array_diff(
    self::get_jetpack_gutenberg_extensions_whitelist(),
    get_option( 'jetpack_extensions_exclusions', array() )
);
```

oikeme replied:
Good catch and thanks. Have updated this to just return the diff

oikeme committed:
Update get_available_exgtensions

Update get_available_extensions to include param for whitelisted extensions or default if null. Updated tests for get_available_extensions

oikeme committed:
Udate Options

Adding new option to get_all_wp_options() in class.jetpack-options.php and Jetpack_Sync_Defaults class. Also removing unused variable in place of returning diff

jeherve reviewed (dismissed):
Tests seem to be failing right now:

```
WP_Test_Jetpack_Sync_Options::test_sync_default_options
Some whitelisted options don't have a test: Array
(
    [135] => jetpack_extensions_exclusions
)
Failed asserting that false is true.
```

oikeme committed:
Update test_class.jetpack-sync-options.php

Adding new option for jetpack_extensions_exclusions to sync options test

ebinnion commented on class.jetpack-options.php:
```diff
@@ -552,6 +552,7 @@ static function get_all_wp_options() {
 			'jpo_site_type',
 			'jpo_homepage_format',
 			'jpo_contact_page',
+			'get_all_wp_options',
```
Correct location, but this should be `jetpack_extensions_exclusions`.

ebinnion committed:
Options: Correct option name to jetpack_extensions_exclusions

ebinnion commented:
If you're fine with my last commit, this LGTM.

ebinnion reviewed (dismissed):
If you're fine with my last commit, this LGTM.

oikeme committed:
Update class.jetpack-options.php

updating option name for jetpack_extensions_exclusions

oikeme committed:
Merge branch 'whitelist_exclusions' of github.com:Automattic/jetpack into whitelist_exclusions

tyxla commented on class.jetpack-gutenberg.php:
```diff
@@ -326,6 +326,20 @@ public static function get_jetpack_gutenberg_extensions_whitelist() {
+	 *
+	 * @return array A list of blocks: eg [ 'publicize', 'markdown' ]
+	 */
+	public static function get_available_extensions( $whitelisted_extensions = null ) {
```
Any particular reason to prefer `null` than `array()` here? If we had `array()` we could:
* Check for emptiness easier - an empty array evaluates as `false`, and a non-empty one evaluates as true.
* Be more specific about the type of data we're expecting here.
* Make sure that we're always using the same data type for this argument.

```suggestion
	public static function get_available_extensions( $whitelisted_extensions = array() ) {
```

ebinnion replied:
Yes. While making `$whitelisted_extensions` be an `array` does make some sense so that it's the same type of data, that also presents us an issue. If we were to change `$whitelisted_extensions` to default to `array`, then we have no way to tell why the array is empty. For example, is it empty because a test forced it to be empty, in which case we should populate it with data. Or, is it empty because that's the default value, in which case we should populate it with data.

tyxla replied:
Thanks for clarifying - makes sense to me 👍

tyxla commented on class.jetpack-gutenberg.php:
```diff
@@ -326,6 +326,20 @@ public static function get_jetpack_gutenberg_extensions_whitelist() {
+	 */
+	public static function get_available_extensions( $whitelisted_extensions = null ) {
+		$exclusions             = get_option( 'jetpack_extensions_exclusions', array() );
+		$whitelisted_extensions = is_null( $whitelisted_extensions ) ? self::get_jetpack_gutenberg_extensions_whitelist() : $whitelisted_extensions;
```
This is the line that could benefit from using an `array()` default value for the `whitelisted_extensions` argument.

```suggestion
		$whitelisted_extensions = ! $whitelisted_extensions ? self::get_jetpack_gutenberg_extensions_whitelist() : $whitelisted_extensions;
```

Alternatively, we could do it like this, which arguably demonstrates the intention better:

```
if( ! $whitelisted_extensions ) {
	$whitelisted_extensions = self::get_jetpack_gutenberg_extensions_whitelist();
}
```

ebinnion replied:
Agreed that the full `if` could be better. Though, I still think the `is_null` check for a default `null` value is a bit better.

tyxla commented on class.jetpack-options.php:
```diff
@@ -552,6 +552,7 @@ static function get_all_wp_options() {
 			'jpo_site_type',
 			'jpo_homepage_format',
 			'jpo_contact_page',
+			'jetpack_extensions_exclusions',
```
Nit: I'd likely go with `jetpack_excluded_extensions` as this would make the option name more consistent with our naming scheme, which often is `(jetpack_)?(ADJECTIVE_)?(NOUN)`.

Definitely not a requirement and no hard feelings about this one (I assume the option name is already decided), just wanted to share my $0.02. 😉

ebinnion replied:
> I assume the option name is already decided

That sounds reasonable to me as well. Should be a simple fix.

oikeme replied:
I've updated this everywhere to be jetpack_excluded_extensions 🙂

tyxla commented on tests/php/test_class.jetpack-gutenberg.php:
```diff
@@ -129,4 +130,18 @@ function test_plugin_is_not_available_when_not_registered_returns_missing_module
+		$extensions = Jetpack_Gutenberg::get_available_extensions( $this->get_extensions_whitelist() );
+		$this->assertInternalType( 'array', $extensions );
+		$this->assertNotEmpty( $extensions );
+		$this->assertNotContains( 'onion', $extensions );
```
Should we also add a test for setting the option back to an empty array, and verifying that this is recognized properly in the same way as when no option exists?

ebinnion replied:
That's a good question. I'm not sure that it's necessary though since we default the `get_option()` to return `array()` when the option does not exist.

tyxla commented on class.jetpack-gutenberg.php:
```diff
@@ -326,6 +326,20 @@ public static function get_jetpack_gutenberg_extensions_whitelist() {
+	 *
+	 * @return array A list of blocks: eg [ 'publicize', 'markdown' ]
+	 */
+	public static function get_available_extensions( $whitelisted_extensions = null ) {
```
Is the `$whitelisted_extensions` argument still necessary? It doesn't seem to be used anywhere else other than in tests.

tyxla commented on tests/php/test_class.jetpack-gutenberg.php:
```diff
@@ -129,4 +130,18 @@ function test_plugin_is_not_available_when_not_registered_returns_missing_module
 	}
+
+	function test_get_available_extensions() {
+		$extensions = Jetpack_Gutenberg::get_available_extensions( $this->get_extensions_whitelist() );
```
Since that's the default usage, should we also add a test for cases where we don't specify a `$whitelisted_extensions` argument?

ebinnion replied:
That's originally how the test was written. But, the list returned from `get_jetpack_gutenberg_extensions_whitelist()` differs between environments. Specifically, that method seems to return an empty array in Travis while returning a proper list of extensions in our local environments. Ideally, we could just mock the return of `get_jetpack_gutenberg_extensions_whitelist()`, but since these are static methods, we'd need to do that by adding a filter to `get_jetpack_gutenberg_extensions_whitelist()` most likely. So, it seemed like adding an argument to short-circuit calling `get_jetpack_gutenberg_extensions_whitelist()` was a good option.

tyxla reviewed (commented):
Code's looking well 👍

Left some questions and thoughts, nothing blocking I guess. 😉

tyxla committed:
Avoid using short array syntax in docs.

WordPress does not support PHP 5.3+ yet.

Co-Authored-By: jeherve <jeremy@tagada.hu>

oikeme committed:
Renaming option

Renaming option for jetpack_extensions_exclusions to jetpack_excluded_extensions

kraftbj commented on tests/php/sync/test_class.jetpack-sync-options.php:
```diff
@@ -201,6 +201,7 @@ public function test_sync_default_options() {
 			'site_user_type'                       => wp_json_encode( array( 1 => 'pineapple' ) ),
 			'site_segment'                         => 'pineapple',
 			'site_vertical'                        => 'pineapple',
+			'jetpack_excluded_extensions'        => 'pineapple',
```
nitpicky - lack of alignment

kraftbj replied:
50092c4

kraftbj committed:
Very minor spacing change

kraftbj commented:
Once tests pass, it is ready to merge and have the wp.com fusion diff committed.

oikeme committed:
Merge branch 'master' into whitelist_exclusions
